src/dyncfg-ld/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+use mz_dyncfg::{ConfigSet, ConfigUpdates, ConfigVal};
+use mz_ore::cast::CastLossy;
+use mz_ore::future::TimeoutError;
+use mz_ore::task;
+use tokio::time;
+
+/// Start a background task that syncs to a ConfigSet from LaunchDarkly. `ctx_builder` should be
+/// used to add additional LD contexts. A `build` context is added automatically. Returns after the
+/// LD client has been initialized or the `config_sync_timeout` timeout has been reached.
+pub async fn sync_ld_to_configset<F>(
+    set: ConfigSet,
+    build_info: &'static BuildInfo,
+    ctx_builder: F,
+    // Use an option so that local dev where this is disabled still uses the same validation logic
+    // for the ConfigSet.
+    launchdarkly_sdk_key: Option<&str>,
+    config_sync_timeout: Duration,
+    config_sync_loop_interval: Option<Duration>,
+) -> Result<(), anyhow::Error>
+where
+    F: FnOnce(&mut ld::MultiContextBuilder),
+{
+    // Ensure that all the ConfigVals in the set support FlagValue conversion, even if LD is
+    // disabled (preventing error skew in local vs prod settings).
+    for entry in set.entries() {
+        let _ = dyn_into_flag(entry.val())?;
+    }
+    let ld_client = if let Some(key) = launchdarkly_sdk_key {
+        let client = ld::Client::build(ld::ConfigBuilder::new(key).build())?;
+        let init = async {
+            let max_backoff = Duration::from_secs(60);
+            let mut backoff = Duration::from_secs(5);
+            while !client.initialized_async().await {
+                tracing::warn!("SyncedConfigSet failed to initialize");
+                tokio::time::sleep(backoff).await;
+                backoff = (backoff * 2).min(max_backoff);
+            }
+            Ok(())
+        };
+        match mz_ore::future::timeout(config_sync_timeout, init).await {
+            Ok(_) => Some(client),
+            Err(TimeoutError::Inner(e)) => return Err(e),
+            Err(TimeoutError::DeadlineElapsed) => {
+                tracing::info!("SyncedConfigSet initialize on boot: initialize has timed out");
+                None
+            }
+        }
+    } else {
+        None
+    };
+
+    let synced = SyncedConfigSet {
+        set,
+        ld_client,
+        ld_ctx: ld_ctx(build_info, ctx_builder)?,
+    };
+    synced.sync()?;
+    task::spawn(
+        || "SyncedConfigSet sync_loop",
+        synced.sync_loop(config_sync_loop_interval),
+    );
+    Ok(())
+}
+
+fn ld_ctx<F>(build_info: &'static BuildInfo, ctx_builder: F) -> Result<ld::Context, anyhow::Error>
+where
+    F: FnOnce(&mut ld::MultiContextBuilder),
+{
+    // Register multiple contexts for this client.
+    //
+    // Unfortunately, it seems that the order in which conflicting targeting
+    // rules are applied depends on the definition order of feature flag
+    // variations rather than on the order in which context are registered with
+    // the multi-context builder.
+    let mut builder = ld::MultiContextBuilder::new();
+
+    builder.add_context(
```

Comment:
If it's easy to share code for this, I'd do it. E.g. make `ld_ctx` pub in this crate and call it from envd. Seems like the type of thing that could easily rot in one codepath.

Reply:
There's already an existing ld_ctx fn in the adapter crate that does specific stuff for an Environment. It's not obviously shareable to me. I think it's fine for the current PR.